File: sdk-python/tests/_conformance/dispatch_forms.py
"""Forms-resource dispatcher for the conformance harness."""

from __future__ import annotations

from typing import Any

import pytest

from threecommon import AsyncThreeCommon, ThreeCommon
from threecommon.forms import (
    AddElementBody,
    AddLogicRuleBody,
    CreateBody,
    DuplicateBody,
    EnableOtherOptionBody,
    ListParams,
    MoveElementBody,
    UpdateBody,
    UpdateElementBody,
)


def build_list_params(args: dict[str, Any]) -> ListParams | None:
    if not args:
        return None
    payload: dict[str, Any] = {}
    mapping = {"page": "page", "pageSize": "page_size", "type": "type"}
    for camel, snake in mapping.items():
        if camel in args:
            payload[snake] = args[camel]
        elif snake in args:
            payload[snake] = args[snake]
    return ListParams.model_validate(payload) if payload else None


def dispatch_sync(client: ThreeCommon, method: str, args: dict[str, Any]) -> Any:  # noqa: ANN401, PLR0911, PLR0912
    if method == "list":
        return client.forms.list(build_list_params(args))
    if method == "listAutoPaginate":
        return list(client.forms.list_auto_paginate(build_list_params(args)))
    if method == "retrieve":
        return client.forms.retrieve(args["id"])
    if method == "create":
        return client.forms.create(CreateBody.model_validate(args.get("body") or {}))
    if method == "update":
        return client.forms.update(args["id"], UpdateBody.model_validate(args.get("body") or {}))
    if method == "duplicate":
        return client.forms.duplicate(
            args["id"], DuplicateBody.model_validate(args.get("body") or {})
        )
    if method == "addElement":
        return client.forms.add_element(
            args["id"], AddElementBody.model_validate(args.get("body") or {})
        )
    if method == "updateElement":
        return client.forms.update_element(
            args["id"], args["elementId"], UpdateElementBody.model_validate(args.get("body") or {})
        )
    if method == "deleteElement":
        return client.forms.delete_element(args["id"], args["elementId"])
    if method == "moveElement":
        return client.forms.move_element(
            args["id"], args["elementId"], MoveElementBody.model_validate(args.get("body") or {})
        )
    if method == "addLogicRule":
        return client.forms.add_logic_rule(
            args["id"], args["elementId"], AddLogicRuleBody.model_validate(args.get("body") or {})
        )
    if method == "removeLogicRule":
        return client.forms.remove_logic_rule(
            args["id"], args["elementId"], args["targetElementId"]
        )
    if method == "enableOtherOption":
        return client.forms.enable_other_option(
            args["id"],
            args["elementId"],
            EnableOtherOptionBody.model_validate(args.get("body") or {}),
        )
    if method == "disableOtherOption":
        return client.forms.disable_other_option(args["id"], args["elementId"])
    pytest.fail(f"unsupported forms method: {method}")


async def dispatch_async(  # noqa: PLR0911, PLR0912
    client: AsyncThreeCommon, method: str, args: dict[str, Any]
) -> Any:  # noqa: ANN401
    if method == "list":
        return await client.forms.list(build_list_params(args))
    if method == "listAutoPaginate":
        return [f async for f in client.forms.list_auto_paginate(build_list_params(args))]
    if method == "retrieve":
        return await client.forms.retrieve(args["id"])
    if method == "create":
        return await client.forms.create(CreateBody.model_validate(args.get("body") or {}))
    if method == "update":
        return await client.forms.update(
            args["id"], UpdateBody.model_validate(args.get("body") or {})
        )
    if method == "duplicate":
        return await client.forms.duplicate(
            args["id"], DuplicateBody.model_validate(args.get("body") or {})
        )
    if method == "addElement":
        return await client.forms.add_element(
            args["id"], AddElementBody.model_validate(args.get("body") or {})
        )
    if method == "updateElement":
        return await client.forms.update_element(
            args["id"], args["elementId"], UpdateElementBody.model_validate(args.get("body") or {})
        )
    if method == "deleteElement":
        return await client.forms.delete_element(args["id"], args["elementId"])
    if method == "moveElement":
        return await client.forms.move_element(
            args["id"], args["elementId"], MoveElementBody.model_validate(args.get("body") or {})
        )
    if method == "addLogicRule":
        return await client.forms.add_logic_rule(
            args["id"], args["elementId"], AddLogicRuleBody.model_validate(args.get("body") or {})
        )
    if method == "removeLogicRule":
        return await client.forms.remove_logic_rule(
            args["id"], args["elementId"], args["targetElementId"]
        )
    if method == "enableOtherOption":
        return await client.forms.enable_other_option(
            args["id"],
            args["elementId"],
            EnableOtherOptionBody.model_validate(args.get("body") or {}),
        )
    if method == "disableOtherOption":
        return await client.forms.disable_other_option(args["id"], args["elementId"])
    pytest.fail(f"unsupported forms method: {method}")

File: sdk-node/test/conformance/dispatch-forms.ts
/**
 * Forms-resource dispatcher for the conformance harness. Kept in its own file
 * so adding new resources doesn't bloat the shared runner.
 */

import { expectBody, expectString, type ScenarioCall } from './scenario'

import type { ThreeCommon } from '@/client'

export function dispatchForms(
  client: ThreeCommon,
  call: ScenarioCall,
): Promise<unknown> | AsyncIterableIterator<unknown> {
  const args = call.args
  switch (call.method) {
    case 'list':
      return client.forms.list(args)
    case 'retrieve':
      return client.forms.retrieve(expectString(args['id'], 'retrieve'))
    case 'create': {
      const body = expectBody(args['body'], 'create')
      return client.forms.create(body as unknown as Parameters<typeof client.forms.create>[0])
    }
    case 'update': {
      const id = expectString(args['id'], 'update')
      const body = expectBody(args['body'], 'update')
      return client.forms.update(id, body)
    }
    case 'duplicate': {
      const id = expectString(args['id'], 'duplicate')
      const body = args['body'] === undefined ? undefined : expectBody(args['body'], 'duplicate')
      return client.forms.duplicate(id, body)
    }
    case 'addElement': {
      const id = expectString(args['id'], 'addElement')
      const body = expectBody(args['body'], 'addElement')
      return client.forms.addElement(
        id,
        body as unknown as Parameters<typeof client.forms.addElement>[1],
      )
    }
    case 'updateElement': {
      const id = expectString(args['id'], 'updateElement')
      const elementId = expectString(args['elementId'], 'updateElement')
      const body = expectBody(args['body'], 'updateElement')
      return client.forms.updateElement(id, elementId, body)
    }
    case 'deleteElement': {
      const id = expectString(args['id'], 'deleteElement')
      const elementId = expectString(args['elementId'], 'deleteElement')
      return client.forms.deleteElement(id, elementId)
    }
    case 'moveElement': {
      const id = expectString(args['id'], 'moveElement')
      const elementId = expectString(args['elementId'], 'moveElement')
      const body = expectBody(args['body'], 'moveElement')
      return client.forms.moveElement(
        id,
        elementId,
        body as unknown as Parameters<typeof client.forms.moveElement>[2],
      )
    }
    case 'enableOtherOption': {
      const id = expectString(args['id'], 'enableOtherOption')
      const elementId = expectString(args['elementId'], 'enableOtherOption')
      const body = expectBody(args['body'], 'enableOtherOption')
      return client.forms.enableOtherOption(
        id,
        elementId,
        body as unknown as Parameters<typeof client.forms.enableOtherOption>[2],
      )
    }
    case 'disableOtherOption': {
      const id = expectString(args['id'], 'disableOtherOption')
      const elementId = expectString(args['elementId'], 'disableOtherOption')
      return client.forms.disableOtherOption(id, elementId)
    }
    case 'addLogicRule': {
      const id = expectString(args['id'], 'addLogicRule')
      const elementId = expectString(args['elementId'], 'addLogicRule')
      const body = expectBody(args['body'], 'addLogicRule')
      return client.forms.addLogicRule(
        id,
        elementId,
        body as unknown as Parameters<typeof client.forms.addLogicRule>[2],
      )
    }
    case 'removeLogicRule': {
      const id = expectString(args['id'], 'removeLogicRule')
      const elementId = expectString(args['elementId'], 'removeLogicRule')
      const targetElementId = expectString(args['targetElementId'], 'removeLogicRule')
      return client.forms.removeLogicRule(id, elementId, targetElementId)
    }
    case 'listAutoPaginate':
      return client.forms.listAutoPaginate(args)
    case 'finalize':
    case 'void':
    case 'recordPayment':
    case 'autoCharge':
    case 'refundPayment':
    case 'deleteDraft':
    case 'activate':
    case 'cancel':
    case 'cancelImmediately':
    case 'markUnpaid':
    case 'bill':
    case 'renew':
    case 'previewUpcomingInvoice':
    case 'count':
    case 'delete':
    case 'bulkUpsert':
    case 'listActivity':
    case 'listActivityAutoPaginate':
    case 'lookup':
    case 'grant':
    case 'consume':
    case 'archive':
    case 'unarchive':
    case 'resolve':
      throw new Error(`forms: unsupported method '${call.method}'`)
  }
}
